shared/src/price_estimation/priority.rs

```diff
@@ -0,0 +1,117 @@
+use super::{Estimate, PriceEstimating, PriceEstimationError, Query};
+use anyhow::Result;
+
+/// Tries inner price estimators in order for queries failing with PriceEstimationError::Other.
+/// Successes, UnsupportedToken or NoLiquidity errors are not retried.
+pub struct PriorityPriceEstimator {
+    inner: Vec<Box<dyn PriceEstimating>>,
+}
+
+impl PriorityPriceEstimator {
+    pub fn new(inner: Vec<Box<dyn PriceEstimating>>) -> Self {
+        assert!(!inner.is_empty());
+        Self { inner }
+    }
+}
+
+#[async_trait::async_trait]
+impl PriceEstimating for PriorityPriceEstimator {
+    async fn estimates(&self, queries: &[Query]) -> Vec<Result<Estimate, PriceEstimationError>> {
+        let mut results = self.inner[0].estimates(queries).await;
```

Comment:
Cleaver. It only retries failed queries instead of an "all or nothing" approach. :+1: